client/network/src/request_responses.rs

```diff
@@ -345,148 +352,185 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
 			Self::OutEvent,
 		>,
 	> {
-		'poll_all: loop {
-			// Poll to see if any response is ready to be sent back.
-			while let Poll::Ready(Some(result)) = self.pending_responses.poll_next_unpin(cx) {
-				match result {
-					RequestProcessingOutcome::Response {
-						protocol, inner_channel, response
-					} => {
-						if let Some((protocol, _)) = self.protocols.get_mut(&*protocol) {
-							protocol.send_response(inner_channel, Ok(response));
+		// Poll to see if any response is ready to be sent back.
+		while let Poll::Ready(Some(result)) = self.pending_responses.poll_next_unpin(cx) {
+			match result {
+				RequestProcessingOutcome::Response {
+					request_id, peer, protocol: protocol_name, inner_channel, response
+				} => {
+					if let Some((protocol, _)) = self.protocols.get_mut(&*protocol_name) {
+						if let Err(_) = protocol.send_response(inner_channel, Ok(response)) {
+							self.pending_responses_arrival_time.pop(&request_id);
+							log::warn!(
+								target: "sub-libp2p",
+								"Failed to send response for {:?} on protocol {:?}. Dropping \
+								 response",
+								request_id, protocol_name,
+							);
+							let out = Event::InboundRequest {
+								peer,
+								protocol: protocol_name,
+								result: Err(ResponseFailure::TimeoutOrClosed),
```

Comment:
Good point, thanks for mentioning this! I agree that we should not couple the general control flow to the `pending_response_arrival_time` cache. I removed the logic emitting an event with https://github.com/paritytech/substrate/pull/7696/commits/bf0b186a158ea56ec034ee96d3dd5e0c46c3f4db.

Reply:
:+1: Seeing the trouble with the `pending_response_arrival_time` here (and the initial "cleanup code") makes me wonder though if we should not better go the extra mile and make sure `libp2p-request-response` emit `InboundFailure::ConnectionClosed`, like for outbound requests, doing the necessary tracking of inbound requests there, so that as a consumer there are no edge-cases where you only know that an inbound request failed due to the _absence of_ a `ResponseSent` event, which is not really something one can easily work with. I always considered handling inbound requests as a secondary, passive thing to do, where the code isn't really interested in getting exhaustive failure or success events. In the sense that, if you want to know if inbound requests are successful, you look at the node that send them, i.e. at the outbound requests of the other node. Thus if you monitor all outbound requests and their success or failure, you know all you want to know. But that of course won't include things like the local processing time of inbound requests, which is what you want to track here.

Reply:
> In the sense that, if you want to know if inbound requests are successful, you look at the node that send them, i.e. at the outbound requests of the other node. Thus if you monitor all outbound requests and their success or failure, you know all you want to know.

I think it is especially important to monitor incoming requests in an environment where one might not control the sending entity, e.g. on Polkadot.

> Seeing the trouble with the pending_response_arrival_time here (and the initial "cleanup code") makes me wonder though if we should not better go the extra mile and make sure libp2p-request-response emits InboundFailure::ConnectionClosed, like for outbound requests, doing the necessary tracking of inbound requests there, so that as a consumer there are no edge-cases where you only know that an inbound request failed due to the absence of a ResponseSent event, which is not really something one can easily work with.

Removing the edge-case on the lower libp2p layer and thus reducing the complexity needed on the libp2p user side sounds good to me. I created https://github.com/libp2p/rust-libp2p/issues/1882 to track this. I hope to get around to tackling this in the future.